diesel/src/query_builder/select_by_statement/dsl_impls.rs

```diff
@@ -0,0 +1,365 @@
+// use super::BoxedSelectByStatement;
+use crate::associations::HasTable;
+use crate::deserialize::TableQueryable;
+// use crate::backend::Backend;
+// use crate::expression::nullable::Nullable;
+use crate::expression::*;
+use crate::insertable::Insertable;
+use crate::query_builder::insert_statement::InsertFromSelect;
+use crate::query_builder::{Query, SelectByStatement, SelectQuery};
+// use crate::query_dsl::boxed_dsl::BoxedDsl;
+use crate::query_dsl::methods::*;
+use crate::query_dsl::*;
+// use crate::query_source::joins::{Join, JoinOn, JoinTo};
+use crate::query_source::joins::JoinTo;
+// use crate::query_source::QuerySource;
+use crate::sql_types::Bool;
+
+impl<S, STMT, Rhs, Kind, On> InternalJoinDsl<Rhs, Kind, On> for SelectByStatement<S, STMT>
+where
+    STMT: InternalJoinDsl<Rhs, Kind, On>,
+{
+    type Output = STMT::Output;
+
+    fn join(self, rhs: Rhs, kind: Kind, on: On) -> Self::Output {
+        self.inner.join(rhs, kind, on)
```

Comment:
Is there any reason not to reapply the already existing select here? So return `SelectByStatement<S, STMT::Output>`?

Reply:
See the other comment, the resolution would be
1. has another maker `SelectByIndex` and change the wildcard `LoadQuery` implement depends on `SelectByIndex`.
2. looser the `Output` of `InternalJoinDsl` not depends on `AsQuery`
3. leave it as is until the [specialization](https://doc.rust-lang.org/unstable-book/language-features/specialization.html) stable.